extensions/amp-consent/0.1/amp-consent.js

```diff
@@ -18,34 +18,199 @@ import {CSS} from '../../../build/amp-consent-0.1.css';
 import {ConsentPolicyManager} from './consent-policy-manager';
 import {ConsentStateManager} from './consent-state-manager';
 import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {assertHttpsUrl} from '../../../src/url';
+import {
+  childElementsByTag,
+  isJsonScriptTag,
+} from '../../../src/dom';
+import {dict} from '../../../src/utils/object';
+import {getServicePromiseForDoc} from '../../../src/service';
 import {isExperimentOn} from '../../../src/experiments';
+import {parseJson} from '../../../src/json';
+import {user} from '../../../src/log';
 
 const CONSENT_STATE_MANAGER = 'consentStateManager';
 const CONSENT_POLICY_MANGER = 'consentPolicyManager';
 const AMP_CONSENT_EXPERIMENT = 'amp-consent';
+const TAG = 'amp-consent';
 
 
 export class AmpConsent extends AMP.BaseElement {
   constructor(element) {
     super(element);
 
-    this.ampdoc = null;
+    /** @private {?./consent-state-manager.ConsentStateManager} */
+    this.consentStateManager_ = null;
 
-    this.checkConsentHref_ = null;
+    /** @private {?./consent-policy-manager.ConsentPolicyManager} */
+    this.consentPolicyManager_ = null;
 
-    this.consentStateManager_ = null;
+    /** @private {!JsonObject} */
+    this.consentConfig_ = dict();
+
+    /** @private {!JsonObject} */
+    this.policyConfig_ = dict();
+
+    /** @private {!Object} */
+    this.consentUIRequired_ = {};
   }
 
   buildCallback() {
-    if (!isExperimentOn(this.ampdoc.win, AMP_CONSENT_EXPERIMENT)) {
+    if (!isExperimentOn(this.win, AMP_CONSENT_EXPERIMENT)) {
       return;
     }
+
+    // TODO: Decide what to do with incorrect configuration.
+    this.assertAndParseConfig_();
+
+    getServicePromiseForDoc(this.getAmpDoc(), CONSENT_POLICY_MANGER)
+        .then(manager => {
+          this.consentPolicyManager_ = manager;
+          this.generateDefaultPolicy_();
+          const policyKeys = Object.keys(this.policyConfig_);
+          for (let i = 0; i < policyKeys.length; i++) {
+            this.consentPolicyManager_.registerConsentPolicyInstance(
+                policyKeys[i], this.policyConfig_[policyKeys[i]]);
+          }
+        });
+
+    getServicePromiseForDoc(this.getAmpDoc(), CONSENT_STATE_MANAGER)
+        .then(manager => {
+          this.consentStateManager_ = manager;
+          this.init_();
+        });
+
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout == Layout.NODISPLAY;
   }
+
+  /**
+   * Init the amp-consent by registering and initiate consent instance.
+   */
+  init_() {
+    const instanceKeys = Object.keys(this.consentConfig_);
+    for (let i = 0; i < instanceKeys.length; i++) {
+      const instanceId = instanceKeys[i];
+      this.consentStateManager_.registerConsentInstance(instanceId);
```

Comment:
`consentStateManager_` may not have resolved yet, making this throw.

Reply:
`init_()` is only called after `getServicePromiseForDoc(this.getAmpDoc(), CONSENT_STATE_MANAGER)` resolved. I don't think that will be the case. Or am I missing anything